design-proposals/network-vhostuser-binding-plugin/network-vhostuser-binding-plugin.md

```diff
@@ -0,0 +1,142 @@
+
+## Definition of Users
+Users of the feature are everyone that deploys a VM.
+
+## User Stories
+- As a user, I want to create a VM with one or serveral `vhostuser` interfaces attached to a userspace dataplane.
+- As a user, I want the `vhostuser` interface to be configured with a specific MAC address.
+- As a user, I want to enable multi-queue on the `vhostuser` interface
+- As a Network Binding Plugin developper, I want the shared socket path to be accessible to virt-launcher pod 
+- As a CNI developper, I want to access the shared vhostuser sockets from Multus pod
+
+## Repos
+Kubevirt repo, and most specificaly cmd/sidecars.
+
+## Design
+This proposal leverages the Kubevirt Network Binding Plugin sidecar framework to implement a new `network-vhostuser-binding`.
+
+`network-vhostuser-binding` role is to implement the modification to the domain XML according to the VMI definition passed through its gRPC service by the `virt-launcher` pod.
+
+`vhostuser` interfaces are defined in the VMI under `spec/domain/devices/interfaces` using the binding name `vhostuser`:
+
+```yaml
+    spec:
+      domain:
+        devices:
+          networkInterfaceMultiqueue: true
+          interfaces:
+          - name: default
+            masquerade: {}
+          - name: net1
+            binding:
+              name: vhostuser
+            macAddress: ca:fe:ca:fe:42:42
+```
+
+`network-vhostuser-binding` translates the VMI definition into libvirt domain XML modifications:
+1. Creates a new interface with `type='vhostuser'`
+2. Set the MAC address if specified in the VMI spec
+3. If `networkInterfaceMultiqueue` is set to `true`, add the number of queues calculated after the number of cores of the VMI
+4. Add `memAccess='shared'` to all NUMA cells elements
+5. Define the device name according to Kubevirt naming schema
+6. Define the `vhostuser` socket path
+
+Below is an example of modified domain XML:
+
+```xml
+<cpu mode="host-model">
+        <topology sockets="2" cores="8" threads="1"></topology>
+        <numa>
+            <cell id="0" cpus="0-7" memory="2097152" unit="KiB" memAccess="shared"/>
+            <cell id="1" cpus="8-15" memory="2097152" unit="KiB" memAccess="shared"/>
+        </numa>
+</cpu>
+<interface type='vhostuser'>
+    <source type='unix' path='/var/run/kubevirt/sockets/poda08a0fcbdea' mode='server'/>
+    <target dev='poda08a0fcbdea'/>
+    <model type='virtio-non-transitional'/>
+    <mac address='ca:fe:ca:fe:42:42'/>
+    <driver name='vhost' queues='8' rx_queue_size='1024' tx_queue_size='1024'/>
+    <alias name='ua-net1'/>
+</interface>
+```
+
+This design leverages the existing `sockets` emptyDir mounted in `/var/run/kubevirt/sockets`. This allows the CNI to bind mount the socket emptyDir (`/var/lib/kubelet/<pod uid>/volumes/kubernetes.io~empty-dir/sockets`) to a host directory available to the dataplane pod through a hostPath mount.
```

Comment:
I'm a bit confused with the changes this proposals has passed through.
Can you please keep the alternatives and the reasons they have been kept out in an appendix?
E.g. https://github.com/kubevirt/community/pull/294#discussion_r1621991963 discussed the DP performing the mounting.

Regarding this intro sentence:
- How exactly can a CNI perform a mount to the pod/container? At the time the CNI is executed, there is not filesystem namespace defined yet. The CNI is also unlikely to have access to the api-server, it is unsafe and limited to trusted binaries only (e.g. Multus).
- The diagram describes in step 4 that `qemu` creates the socket.
  First, it will be nice to see that diagram steps described in details in text as well.
  In general, libvirt handles interfaces in an unmanaged manner, i.e. all it needs is provided to it (e.g. tap device). This is done to allow libvirt to operate in an unprivileged/rootless mode.
  - Is the socket creation possible with the rootless priv?
  - Can the socket be created by something else and then just consumed by libvirt?

Reply:
> Regarding this intro sentence:
> 
> * How exactly can a CNI perform a mount to the pod/container? At the time the CNI is executed, there is not filesystem namespace defined yet. The CNI is also unlikely to have access to the api-server, it is unsafe and limited to trusted binaries only (e.g. Multus).

Indeed CNI needs to access api-server. Multus 3 allowed that through its own kubeconfig.multus. With Multus 4 thick mode, it's  [no longer possible,](https://github.com/k8snetworkplumbingwg/multus-cni/issues/1258#issue-2233364757) and CNI requesting api-server access needs to handle their own credentials creation.

> * The diagram describes in step 4 that `qemu` creates the socket.
>   First, it will be nice to see that diagram steps described in details in text as well.
>   In general, libvirt handles interfaces in an unmanaged manner, i.e. all it needs is provided to it (e.g. tap device). This is done to allow libvirt to operate in an unprivileged/rootless mode.
>   
>   * Is the socket creation possible with the rootless priv?

Yes it is, as far the filesystem permissions AND SElinux allows it.

>   * Can the socket be created by something else and then just consumed by libvirt?

It can. vhostusers sockets are created in server mode, and consumed in client mode. Usually qemu is in server mode, and the data plane the client. By the way the other way is getting [deprecated in ovs-dpdk](https://docs.openvswitch.org/en/latest/topics/dpdk/vhost-user/#vhost-user-vs-vhost-user-client) for example.